base/loading.jl

```diff
@@ -272,3 +297,73 @@ function compile(name::ByteString)
     create_expr_cache(path, cachefile)
     return cachefile
 end
+
+module_uuid(m::Module) = ccall(:jl_module_uuid, UInt64, (Any,), m)
+
+isvalid_cache_header(f::IOStream) = 0 != ccall(:jl_deserialize_verify_header, Cint, (Ptr{Void},), f.ios)
+
+function cache_dependencies(f::IO)
+    modules = Tuple{Symbol,UInt64}[]
+    files = ByteString[]
+    while true
+        n = ntoh(read(f, Int32))
+        n == 0 && break
+        push!(modules,
+              (symbol(readbytes(f, n)), # module symbol
+               ntoh(read(f, UInt64)))) # module UUID (timestamp)
+    end
+    read(f, Int64) # total bytes for file dependencies
+    while true
+        n = ntoh(read(f, Int32))
+        n == 0 && break
+        push!(files, bytestring(readbytes(f, n)))
+    end
+    return modules, files
+end
+
+function cache_dependencies(cachefile::AbstractString)
+    io = open(cachefile, "r")
+    try
+        !isvalid_cache_header(io) && throw(ArgumentError("invalid cache file $cachefile"))
+        return cache_dependencies(io)
+    finally
+        close(io)
+    end
+end
+
+function stale_cachefile(cachefile::AbstractString, cachefile_mtime::Real=mtime(cachefile))
+    io = open(cachefile, "r")
+    try
+        if !isvalid_cache_header(io)
+            return true # invalid cache file
+        end
+        modules, files = cache_dependencies(io)
+        for f in files
+            if mtime(f) > cachefile_mtime
```

Comment:
`cache_dependencies` returns all of the included files, and compiling a `.ji` doesn't modify the `mtime` for each of those included files, so I suspect this would result in recompiling every time.